src/vsr/replica_test.zig

```diff
@@ -0,0 +1,920 @@
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.R0).stop();
+    t.replica(.R0).corrupt(.{ .wal_prepare = 15 });
+    try expectEqual(t.replica(.R0).open(), .WALCorrupt);
+}
+
+test "Cluster: recovery: WAL header corruption (R=1)" {
+    // R=1 locally repairs WAL-header corruption.
+    const t = try TestContext.init(.{ .replica_count = 1 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.R0).stop();
+    t.replica(.R0).corrupt(.{ .wal_header = 15 });
+    try expectEqual(t.replica(.R0).open(), .ok);
+    try expectEqual(c.request(30), 30);
+}
+
+test "Cluster: recovery: WAL torn prepare, standby with intact prepare (R=1 S=1)" {
+    // R=1 recovers to find that its last prepare was a torn write, so it is truncated.
+    // The standby received the prepare, though.
+    //
+    // R=1 handles this by incrementing its view during recovery, so that the standby can truncate
+    // discard the truncated prepare.
+    const t = try TestContext.init(.{
+        .replica_count = 1,
+        .standby_count = 1,
+    });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.R0).corrupt(.{ .wal_header = 20 });
+    t.replica(.R0).stop();
+    try expectEqual(t.replica(.R0).open(), .ok);
+    try expectEqual(c.request(30), 30);
+    try expectEqual(t.replica(.R0).commit(), 30);
+    try expectEqual(t.replica(.S0).commit(), 30);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, symmetric)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .bidirectional);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+    try expectEqual(t.replica(.B1).commit(), 40);
+    try expectEqual(t.replica(.B2).commit(), 20);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, asymmetric, send-only)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .incoming);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+    try expectEqual(t.replica(.B1).commit(), 40);
+    try expectEqual(t.replica(.B2).commit(), 20);
+}
+
+test "Cluster: network: partition 2-1 (isolate backup, asymmetric, receive-only)" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    t.replica(.B2).drop_all(.__, .outgoing);
+    try expectEqual(c.request(40), 40);
+    try expectEqual(t.replica(.A0).commit(), 40);
+}
```

Comment:
I think depending on the seed, B2 might actually be able to commit 1 or 2 more ops, so I will put
```
    try expect(t.replica(.B2).commit() >= 20);
```
instead of a precise `expectEqual`.